### ISSUE TITLE
All IBC Scenarios Are Treated as Duplicates

### ISSUE THREAD

**Issue by brianrob** [dotnet/coreclr#24093]
Per @briansull, it appears that all IBC scenarios on Linux are treated as duplicates because the command line is only returning the exe name and not the full command line.

This results in only a single process across all of the scenarios being used to optimize the target assembly.

Linux:
```
    <Scenario ordinal="1" scenario="A" mask="0x0001" priority="0" numRuns="5" name="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet">
      <Run runTime="1/1/1601 1:00:00" cmdLine="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet" systemInfo="&lt;machine,OS&gt;" mvid="b1e0f044-1b92-4e7f-b5fe-90220dba9938" />
      <Run runTime="1/1/1601 1:00:00" cmdLine="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet" systemInfo="&lt;machine,OS&gt;" mvid="b1e0f044-1b92-4e7f-b5fe-90220dba9938" />
     <Run runTime="1/1/1601 1:00:00" cmdLine="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet" systemInfo="&lt;machine,OS&gt;" mvid="b1e0f044-1b92-4e7f-b5fe-90220dba9938" />
      <Run runTime="1/1/1601 1:00:00" cmdLine="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet" systemInfo="&lt;machine,OS&gt;" mvid="b1e0f044-1b92-4e7f-b5fe-90220dba9938" />
      <Run runTime="1/1/1601 1:00:00" cmdLine="/__w/1/s/artifacts/output/FXx64LINmas/IBC/cli/x64/dotnet" systemInfo="&lt;machine,OS&gt;" mvid="b1e0f044-1b92-4e7f-b5fe-90220dba9938" />
    </Scenario>
```
Windows:
```
    <Scenario ordinal="1" scenario="A" mask="0x0008" priority="0" numRuns="1" name="dotnet.exe">
      <Run runTime="3/13/2019 19:48:05" cmdLine="&quot;dotnet.exe&quot; exec F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\MSBuild.dll -maxcpucount -verbosity:m -nologo -target:Restore F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\DotNet_HelloWorld\DotNet_HelloWorld.csproj -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll" systemInfo="&lt;machine,OS&gt;" mvid="c2b7bd8d-4c01-4d7a-bf5e-bc3538175ef0" />
    </Scenario>
    <Scenario ordinal="2" scenario="B" mask="0x0004" priority="0" numRuns="1" name="dotnet.exe">
      <Run runTime="3/13/2019 19:48:24" cmdLine="&quot;dotnet.exe&quot; exec F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\MSBuild.dll -maxcpucount -verbosity:m -nologo -target:Restore -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll" systemInfo="&lt;machine,OS&gt;" mvid="c2b7bd8d-4c01-4d7a-bf5e-bc3538175ef0" />
    </Scenario>
    <Scenario ordinal="3" scenario="C" mask="0x0002" priority="0" numRuns="1" name="F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\dotnet.exe">
      <Run runTime="3/13/2019 19:48:42" cmdLine="&quot;F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\dotnet.exe&quot;  exec &quot;F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\Roslyn\bincore\csc.dll&quot; /noconfig @&quot;C:\Users\vsagent\AppData\Local\Temp\tmp91b053281ab444bc847231338352eda8.rsp&quot;" systemInfo="&lt;machine,OS&gt;" mvid="c2b7bd8d-4c01-4d7a-bf5e-bc3538175ef0" />
    </Scenario>
    <Scenario ordinal="4" scenario="D" mask="0x0001" priority="0" numRuns="1" name="dotnet.exe">
     <Run runTime="3/13/2019 19:48:30" cmdLine="&quot;dotnet.exe&quot; exec F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\MSBuild.dll -maxcpucount -verbosity:m -restore -consoleloggerparameters:Summary -target:Build -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,F:\vsagent\51\s\artifacts\output\FXx64WINmas\IBC\cli\x64\sdk\3.0.100-preview4-010738\dotnet.dll" systemInfo="&lt;machine,OS&gt;" mvid="c2b7bd8d-4c01-4d7a-bf5e-bc3538175ef0" />
    </Scenario>
```

**Comment by briansull** [comment]
On Linux the method  `PAL_InitializeCoreCLR` creates a fake command line with only the `ExePath`:

```
    // Fake up a command line to call PAL initialization with.
    int result = Initialize(1, &szExePath, PAL_INITIALIZE_CORECLR);
```

**Comment by briansull** [comment]
When running IBCtraining scenario's we populate a  `CORBBTPROF_SCENARIO_RUN `

```
    //
    // Allocate and initialize the scenario header section
    //
    {
        LPCWSTR  pCmdLine    = GetCommandLineW();
        S_SIZE_T cCmdLine = S_SIZE_T(wcslen(pCmdLine));

        //
        // Allocate the Scenario Run struct
        // 
        {
            CORBBTPROF_SCENARIO_RUN *sRun;
            sRun = (CORBBTPROF_SCENARIO_RUN *)  profileMap->Allocate(sizeRun.Value());
            
            sRun->runTime     = runTime;
            sRun->mvid        = *pMvid;
            sRun->cCmdLine    = cCmdLine.Value();
            sRun->cSystemInfo = cSystemInfo.Value();
            wcscpy_s(sRun->cmdLine, cCmdLine.Value(), pCmdLine);
            wcscpy_s(sRun->cmdLine+cCmdLine.Value(), cSystemInfo.Value(), pSystemInfo);
        }
```

**Comment by briansull** [comment]
The best fix appears to be using `procfs` on Linux systems to retrieve the original command line.

https://en.wikipedia.org/wiki/Procfs

**Comment by briansull** [comment]
second issue is to convert the date-time stamp value returned by `GetProcessTimes` to the Windows form:
IbcData:     `<Run runTime="12/31/1600 4:00:01 PM" `

```
        FILETIME runTime, unused1, unused2, unused3;
        GetProcessTimes(GetCurrentProcess(), &runTime, &unused1, &unused2, &unused3);
```

**Comment by brianrob** [comment]
The command line does get setup properly elsewhere because managed code needs access to it.  EventPipe consumes the proper command line as well and logs it in traces.  Take a look at https://github.com/dotnet/coreclr/blob/fdc9998609263d9f531079e2cbe0f3f31eba3382/src/vm/corhost.cpp#L336.

**Comment by briansull** [comment]
I did take a look at that as well.
That  sets up the argc, argv[] as arguments for calling the main entry point of the managed code, 
with the value of pwzAssemblyPath being what we want as argv[0], or arg[1] if we have corerun.exe as argv[0].

It is best to simply get the original command line using `procfs`

**Comment by brianrob** [comment]
If you want to use procfs, I think that is fine.  I believe corefx does this as well in some places.

**Comment by jkotas** [comment]
procfs non portable. I would just use the managed command line, like event pipe.

**Comment by briansull** [comment]
PR #24229 addresses this issue

**Comment by briansull** [comment]
Resolved with #24283
